### PR TITLE
Fixes connect_loc related hard dels

### PR DESCRIPTION
**Pull request by ninjanomnom**
Probably nothing visible to players but some hard dels were happening in regards to connect_loc when their attached object was being qdeleted and not manually removing the element.

Not 100% sure this fixes them all but the dels that existed with edibles and airlocks were cleared up by these changes.

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
I see the same error on my pr #58916

**Comment by LemonInTheDark** [comment]
I don't think `ELEMENT_DETACH` and `ELEMENT_COMPLEX_DETACH` should be allowed on the same element. One implies that you can just not delete the element on the parent's Destroy() and it'll be all good, and the other says that element.detach() REQUIRES extra context. 

It seems foolish to have both of them attached to the same thing, since `ELEMENT_DETACH` leads maintainers to believe that not calling RemoveElement is ok, when even with this change it has higher overhead attached to it, and might cause anomalous behavior.

**Comment by ninjanomnom** [comment]
All element detach does is call detach when the object attached to is qdeleted. Complex detach doesn't imply that you *require* those additional arguments to detach and as shown here, you don't; it's just a way to specify exactly what you want detached and if no tracked object is specified then everything for that listener should be detached instead. Not calling RemoveElement *is* okay and this pr allows you to safely just qdel the object the element is attached to. I don't want to allow situations where an element needs to be detached "properly".

**Comment by LemonInTheDark** [comment]
I can understand that, a slightly slower qdel is much more preferable then a harddelete like the ones we saw here. It's still best practice to manually remove the element tho isn't it?

File: code/datums/elements/connect_loc.dm [hunk 1]
```diff
@@ -54,6 +52,15 @@
...
 		RegisterSignal(tracked.loc, COMSIG_TURF_CHANGE, .proc/on_turf_change)
 
+/datum/element/connect_loc/proc/unregister_all(datum/listener)
+	for(var/atom/location as anything in targets)
+		var/list/loc_targets = targets[location]
+		for(var/atom/movable/tracked as anything in loc_targets)
+			if(loc_targets[tracked] != listener)
+				continue
+			unregister_signals(listener, tracked, location)
+			UnregisterSignal(tracked, COMSIG_MOVABLE_MOVED)
```
Comment (LemonInTheDark) [inline code/datums/elements/connect_loc.dm:62]: Is this ok if there's more then one thing hooked into the listener?

Comment (ninjanomnom) [reply code/datums/elements/connect_loc.dm:62]: I was thinking about this a bit as I was working on this and I suspect there's probably a bug in here involving that, but this is how it worked before and I'd want to fix this issue separately if it is a problem.

Comment (LemonInTheDark) [reply code/datums/elements/connect_loc.dm:62]: what do you mean by worked before? 
Oh also you may want to add a if(tracked) to that unregistersignal line, since it'll runtime for the valid location of null

Comment (ninjanomnom) [reply code/datums/elements/connect_loc.dm:62]: Look in Detach, the signal was unregistered from the tracked object and didn't do any checks there either. There probably is a problem there but it'll have to be looked at in more detail in another pr

File: code/datums/elements/connect_loc.dm [hunk 2]
```diff
@@ -24,25 +24,23 @@
...
-	if(!istype(tracked))
-		return
-
-	unregister_signals(listener, tracked, tracked.loc)
-
-	UnregisterSignal(tracked, COMSIG_MOVABLE_LOCATION_CHANGE)
+		unregister_all(listener)
+	else if(targets[tracked.loc]) // Detach can happen multiple times due to qdel
+		unregister_signals(listener, tracked, tracked.loc)
+		UnregisterSignal(tracked, COMSIG_MOVABLE_MOVED)
```
Comment (LemonInTheDark) [inline code/datums/elements/connect_loc.dm:37]: This is dead code isn't it? since the line above it should always be run by Destroy()?

**Comment by LemonInTheDark** [comment]
image
I'm not an idiot don't worry about it please don't fire me mr headcoder

File: code/datums/elements/connect_loc.dm (hunk 2 above)
Comment (Mothblocks) [reply code/datums/elements/connect_loc.dm:37]: No, Detach can run without Destroy, and it needs to clean itself up properly.

Comment (Mothblocks) [reply code/datums/elements/connect_loc.dm:37]: Oh wait I'm dumb, you might be right.

Comment (LemonInTheDark) [reply code/datums/elements/connect_loc.dm:37]: it's dead in that nothing currently uses it, but it could be used in theory so it's not kill

**Comment by LemonInTheDark** [comment]
god I give up, I'm just gonna ask ninja to reset his branch in the morning. The ref tracking stuff leaked in from my testing branch, and I can't god damn get the pr's branch to reset, no matter how much I push to it. hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

This is the commit history it ought to be using, that's what it looked like before I got my grubby hands on it + my fixes https://github.com/LemonInTheDark/tgstation/tree/hfasd

**Comment by LemonInTheDark** [comment]
I'm so sorry
